gustaf/ffd.py

```diff
@@ -1,4 +1,579 @@
+        #     self._scale = None
+
+        # else:
+        #     self._logd("  Non-partial!")
+
+
+    def _scale_parametric_dimension_to_hypercube(self):
+        """Scales all knot_vectors of the spline to a range of [0,1]
+        """
+        if self._is_parametric_room_hypercube(self._spline):
+            pass
+        else:
+            knv = []
+            for knot_vector in self._spline.knot_vectors:
+                knot_vector = [kv_v - knot_vector[0] for kv_v in knot_vector]
+                knot_vector = [kv_v / knot_vector[-1] for kv_v in knot_vector]
+                knv.append(knot_vector)
+            self._spline.knot_vectors = knv
+
+
+    def _is_parametric_room_hypercube(
+            self,
+            spline: SPLINE_TYPES
+        ) -> bool:
+        """Checks if the parametric range of the spline is a hypercube.
+
+        Parameters
+        ----------
+        spline : SPLINE_TYPES
+            New spline for which to check for hypercubeness
+
+        Returns
+        -------
+        bool
+            True if parametric room is hypercube, else False
+        """
+        if "Bezier" not in spline.whatami:
+            for knot_vector in spline.knot_vectors:
+                # check if knot_vectors fist element is 0 and last element is 1
+                if not (knot_vector[0] == 0 and knot_vector[-1] == 1):
+                    return False
+        return True
+               
+    def _is_mesh_dimension_same_as_spline_parametric_dimension(self) -> bool:
+        """Checks if the mesh and parametric dimension of the spline are the 
+        same.
+
+        Returns
+        -------
+        bool
+            True if the dimensions are the same, else False.
+        """   
+        if self._mesh.vertices.shape[1] == len(self._spline_offset):
+            True
+        False
+
+    def _scale_mesh_vertices(self):
+        """
+        Scales the mesh vertices into the dimension of a hypercube and save 
+        them in self._q_vertices.
+        """        
+        self._logd("Fitting mesh into spline's parametric space.")
+        if self._is_partial:
+            # currently not supported
+            # should add functionality here for it
+            # define the q_vertices here
+            pass
+        else:
+            self._q_vertices = self._mesh.vertices.copy()
+        
+        original_mesh_bounds = self._mesh.get_bounds()
+        self._mesh_offset = list()
+        self._mesh_scale = list()
+        for mesh_bound  in original_mesh_bounds.T:
+            self._mesh_offset.append(-mesh_bound[0])
+            self._mesh_scale.append(1/(mesh_bound[-1]-mesh_bound[0]))
+
+        # transform vertices
+        self._q_vertices += self._mesh_offset
+        # scale vertices
+        scale_matrix = np.eye(len(self._mesh_scale))*self._mesh_scale
+        self._q_vertices = self._q_vertices @ scale_matrix
+        self._logd("Successfully scaled and transformed mesh vertices!")
+
```

Comment:
```
self._mesh_offset = original_mesh_bounds[0]
self._mesh_scale = 1 / (original_mesh_bounds[1] - original_mesh-bounds[0])
# get rid of inf
self._q_vertices *= self._mesh_scale
```

Reply:
Get rid of inf is not treated. I don't see where this would apply, except for the division in case mesh bounds are trivial